src/adapter/src/session.rs

```diff
@@ -1457,3 +1466,181 @@ impl From<&ExplainContext> for RequireLinearization {
+}
+
+impl WriteLocksBuilder {
+    /// Adds a lock to this builder.
+    pub fn insert_lock(&mut self, id: GlobalId, lock: tokio::sync::OwnedMutexGuard<()>) {
+        self.locks.insert(id, Some(lock));
+    }
+
+    /// Finish this builder by returning either all of the necessary locks, or none of them.
+    pub fn all_or_nothing(self, conn_id: &ConnectionId) -> Result<WriteLocks, BTreeSet<GlobalId>> {
+        let (locks, missing): (BTreeMap<_, _>, BTreeSet<_>) =
+            self.locks
+                .into_iter()
+                .partition_map(|(gid, lock)| match lock {
+                    Some(lock) => itertools::Either::Left((gid, lock)),
+                    None => itertools::Either::Right(gid),
+                });
+
+        if missing.is_empty() {
+            tracing::info!(%conn_id, ?locks, "acquired write locks");
+            Ok(WriteLocks {
+                locks,
+                conn_id: conn_id.clone(),
+            })
+        } else {
+            // Explicitly drop the already acquired locks.
+            drop(locks);
+            Err(missing)
+        }
+    }
+}
+
+/// Collection of [`WriteLocks`] gathered during [`group_commit`].
+///
+/// Note: This struct should __never__ be used outside of group commit because it attempts to merge
+/// together several collections of [`WriteLocks`] which if not done carefully can cause deadlocks
+/// or consistency violations.
+///
+/// We must prevent writes from occurring to tables during read then write plans (e.g. `UPDATE`)
+/// but we can allow blind writes (e.g. `INSERT`) to get committed concurrently at the same
+/// timestamp when submitting the updates from a read then write plan.
+///
+/// Naively it would seem as though we could allow blind writes to occur whenever as blind writes
+/// could never cause invalid retractions, but it could cause us to violate serializability because
+/// there is no total order we could define for the transactions. Consider the following scenario:
+///
+/// ```text
+/// table: foo
+///
+///  a | b
+/// --------
+///  x   2
+///  y   3
+///  z   4
+///
+/// -- Session(A)
+/// -- read then write plan, reads at t0, writes at t3, transaction Ta
+/// DELETE FROM foo WHERE b % 2 = 0;
+///
+///
+/// -- Session(B)
+/// -- blind write into foo, writes at t1, transaction Tb
+/// INSERT INTO foo VALUES ('q', 6);
+/// -- select from foo, reads at t2, transaction Tc
+/// SELECT * FROM foo;
+///
+///
+/// The times these operations occur at are ordered:
+/// t0 < t1 < t2 < t3
+///
+/// The initial read for the `DELETE` occurs at t0, so it won't observe ('q', 6) which was inserted
+/// at time t1, thus Ta must be ordered before Tb. But the read at t2 won't observe the deletion of
+/// ('x', 2) and ('z', 4) since the retractions won't get submitted until t3, but it will observe
+/// the insertion of ('q', 6), thus Tc must be ordered before Ta and after Tb. This ordering isn't
+/// possible though, thus no total order exists for these transactions.
```

Comment:
It might be worth explicitly spelling out the partial order (with `<` notation) of the transactions to make it clear that no total order exists. Something like:

- Ta does not read the row with ('q', 6), so Ta < Tb.
- Tc does see the row with ('q', 6), so Tb < Tc.
- Tc does not see the retractions from Ta, so Tc < Ta.

Reply:
For a total order to exist the following must be true, Ta < Tb < Tc < Ta, which is clearly impossible.

Reply:
Good call, updated to use your explanation with the `<` notation which is more clear